CB-3103 Avoid database corruption

EmYar opened the pull request:
Adds savepoints before each update script

EmYar committed:
CB-3103 Avoid database corruption

Adds savepoints before each update script

EmYar committed:
CB-3103 Avoid database corruption

Replaces savepoints by transaction rollback on errors

LonwoLonwo commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/schema/SQLSchemaManager.java:
```diff
@@ -82,29 +82,37 @@ public SQLSchemaManager(
+                    }
+
+                    txn.commit();
+                } catch (Exception e) {
```
How about adding logging for this situation? It can be helpful here, I think.

EmYar committed:
CB-3103 Avoid database corruption

Adds migration rollback logging

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/schema/SQLSchemaManager.java:
```diff
@@ -51,22 +51,22 @@ public final class SQLSchemaManager {
+            SQLDialect targetDatabaseDialect,
+            String targetDatabaseName,
+            int schemaVersionActual,
+            int schemaVersionObsolete
```
Please change the continuation indent from 8 spaces to 4 spaces.

In IntelliJ IDEA, it can be configured in <kbd>Editor</kbd> &rArr; <kbd>Code Style</kbd> &rArr; <kbd>Java</kbd> &rArr; <kbd>Tabs and Indents</kbd> &rArr; <kbd>Continuation indent</kbd>.

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/schema/SQLSchemaManager.java:
```diff
@@ -82,29 +82,38 @@ public SQLSchemaManager(
-                    versionManager.updateCurrentSchemaVersion(monitor, dbCon, targetDatabaseName);
-                } else if (schemaVersionActual > currentSchemaVersion) {
-                    upgradeSchemaVersion(monitor, dbCon, currentSchemaVersion);
+                try {
```
This nested `try` doesn't seem to be necessary? You can add `finally` to the `try-with-resources` block, too.

EmYar replied:
The catch block from the nested `try` uses `txn`, which is not available from the `try-with-resource`'s `catch` block.

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/schema/SQLSchemaManager.java:
```diff
@@ -113,19 +122,19 @@ private void upgradeSchemaVersion(DBRProgressMonitor monitor, Connection connect
-                    throw e;
-                } finally {
-                    ContentUtils.close(ddlStream);
-                }
+            if (ddlStream == null) {
+                continue;
+            }
```
The `try-with-resources` will do better in this situation. It's `null`-aware, so you can safely do the following:

```java
try (Reader ddlStream = scriptSource.openSchemaUpdateScript(monitor, updateToVer)) {
    ...
}

ShadelessFox commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/schema/SQLSchemaVersionManager.java:
```diff
@@ -34,10 +35,24 @@ public interface SQLSchemaVersionManager {
+     * Updates current schema version to actual
+     */
+    void updateCurrentSchemaVersion(
+            DBRProgressMonitor monitor,
+            @NotNull Connection connection,
+            @NotNull String schemaName
+    ) throws DBException, SQLException;
+
     /**
      * Updates current schema version
      */
-    void updateCurrentSchemaVersion(DBRProgressMonitor monitor, Connection connection, String schemaName) throws DBException, SQLException;
+    void updateCurrentSchemaVersion(
+            DBRProgressMonitor monitor,
+            @NotNull Connection connection,
+            @NotNull String schemaName,
+            int version
+    ) throws DBException, SQLException;
```
Same code style issue here

EmYar committed:
Merge branch 'devel' into CB-3103-Avoid-database-corruption

EmYar committed:
CB-3103 Avoid database corruption

Fixes code style issues

EmYar committed:
Merge branch 'devel' into CB-3103-Avoid-database-corruption

EmYar committed:
CB-3103 Avoid database corruption

Removes SQLSchemaVersionManager#updateCurrentSchemaVersion implementation with actual version value by defdault.
Adds SQLSchemaVersionManager#getLatestSchemaVersion to use it in updateCurrentSchemaVersion#updateCurrentSchemaVersion(DBRProgressMonitor, Connection, String, int) when an actual version is needed

EmYar committed:
Merge branch 'devel' into CB-3103-Avoid-database-corruption

EmYar committed:
CB-3103 Avoid database corruption

Fixes code style issues

EmYar committed:
Merge branch 'devel' into CB-3103-Avoid-database-corruption

EmYar committed:
Merge branch 'devel' into CB-3103-Avoid-database-corruption

EmYar committed:
Merge branch 'devel' into CB-3103-Avoid-database-corruption

EmYar committed:
CB-3103
Avoid database corruption

Adds commit after each migration
